### PR TITLE
Faster, non-blocking world init

### PR DESCRIPTION
**Pull request by Cyberboss**
1hydi9

Changes all turf/New calls into Initialize calls (with some optimization for /turf/open/space), moving the hard work off of world startup into SSobj which is much more manageable (based CHECK_TICK)

### CODE REVIEW

File: code/game/turfs/closed.dm [hunk 1]
```diff
@@ -34,19 +34,14 @@
...
 	if(titlescreen)
 		icon_state = titlescreen
 
 /turf/closed/indestructible/riveted
 	icon_state = "riveted"
 
-/turf/closed/indestructible/New()
-	..()
-	if(smooth)
-		queue_smooth(src)
```
Comment (Cyberboss) [inline code/game/turfs/closed.dm]: This isn't necessary, the exact same check/call is in turf/New

File: code/game/turfs/open.dm [hunk 2]
```diff
@@ -40,11 +40,6 @@
 	baseturf = /turf/open/indestructible/hierophant
 	smooth = SMOOTH_TRUE
 
-/turf/open/indestructible/hierophant/New()
-	..()
-	if(smooth)
-		queue_smooth(src)
```
Comment (Cyberboss) [inline code/game/turfs/open.dm]: Same here

File: code/game/gamemodes/wizard/spellbook.dm [hunk 3]
```diff
@@ -568,7 +568,8 @@
 	else
 		user << "It appears to have no author."
 
-/obj/item/weapon/spellbook/proc/Initialize()
+/obj/item/weapon/spellbook/Initialize()
```
Comment (lzimann) [inline code/game/gamemodes/wizard/spellbook.dm]: I don't think there's any spellbook at roundstart(I can't remember now if your New/initialize refactor will always call Init on new()), which means they will end up without spells.

**Comment by Cyberboss** [comment]
Initialize is called on all atoms when SSobj is initialized. After that,
it's called from atom/New.

File: code/game/gamemodes/wizard/spellbook.dm [hunk 4]
```diff
@@ -579,11 +580,6 @@
 			qdel(E)
 	tab = categories[1]
 
-/obj/item/weapon/spellbook/New()
-	..()
-	Initialize()
```
Comment (Cyberboss) [inline code/game/gamemodes/wizard/spellbook.dm]: Also the spellbook was stupid because look at this shit

File: code/game/gamemodes/wizard/spellbook.dm (hunk 3 above)
Comment (Cyberboss) [inline code/game/gamemodes/wizard/spellbook.dm]: Initialize is called on all atoms when SSobj is initialized. After that, it's called from atom/New. Also the spell book was

**Comment by MrStonedOne** [comment]
>This is the only downside: Initialized Objects subsystem within 22 seconds!

Because of how varied that is on people's processors, that number (22 seconds) means literally nothing without a before number.

If you had a slow processor, it could have gone from 15 to 22, but if you have a super fast processor, it could have gone from 3 seconds to 22 seconds.

I'd need to see better numbers to gauge how much this matters.

**Comment by Cyberboss** [comment]
Running a benchmark fully merged based on the previous commit:
i5 3770K Clocked to 4.3GHz
Loading BoxStation

Master:
World init: 28.858s
SSObj: 1.3s

PR:
World init: 18.815s
SSObj: 25.8s

File: code/game/turfs/space/space.dm [hunk 5]
```diff
@@ -15,7 +15,8 @@
 	var/global/datum/gas_mixture/space/space_gas = new
 	plane = PLANE_SPACE
 
-/turf/open/space/New()
+/turf/open/space/Initialize()
+	..()
```
Comment (MrStonedOne) [inline code/game/turfs/space/space.dm]: One thing you can do is not call parent from here.

That should lead to a drop in ss init time.

(The parent call has always been intentionally left out from here.)

**Comment by Cyberboss** [comment]
PR (latest commit):
World init: 9.232s
SSObj: 17.9s

Now we're talking!

**Comment by Cyberboss** [comment]
Wait, atmos runtimes...

**Comment by Cyberboss** [comment]
Same stats

**Comment by MrStonedOne** [comment]
....

That's a net decrease in total world startup time of like 5 seconds.

**Comment by MrStonedOne** [comment]
AND its tick_checked.

So not only is it a net decrease in total startup time, but it also makes startup non-blocking.

**Comment by MrStonedOne** [comment]
Update the OP.

**Comment by MrStonedOne** [comment]
Test merged on sybil

**Comment by Cyberboss** [comment]
Smooth as fucking butter

**Comment by MrStonedOne** [comment]
````
[21:48:49]Maximum recursion level reached (perhaps there is an infinite loop)
To avoid this safety check, set world.loop_checks=0.
[21:48:49]proc name: add to active (/datum/subsystem/air/proc/add_to_active)
[21:48:49]  usr: null
[21:48:49]  src: Air (/datum/subsystem/air)
[21:48:49]  call stack:
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,144,8) (/turf/open/space), null)
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,145,8) (/turf/open/space), 1)
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,144,8) (/turf/open/space), 1)
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,145,8) (/turf/open/space), 1)
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,144,8) (/turf/open/space), 1)
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,145,8) (/turf/open/space), 1)
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,144,8) (/turf/open/space), 1)
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,145,8) (/turf/open/space), 1)
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,144,8) (/turf/open/space), 1)
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,145,8) (/turf/open/space), 1)
[21:48:49]...
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,145,8) (/turf/open/space), 1)
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,145,8) (/turf/open/space), 1)
[21:48:49]Air (/datum/subsystem/air): add to active(space (12,145,8) (/turf/open/space), 1)
[21:48:49]space (12,144,8) (/turf/open/space): air update turf(1)
[21:48:49]the reinforced window (/obj/structure/window/reinforced): New(space (12,144,8) (/turf/open/space), 0)
[21:48:49]/dmm_suite (/dmm_suite): parse grid("/obj/structure/window/reinforc...", 12, 144, 8)
[21:48:49]Kosmicheskaya Stantsiya 13 (/datum/map_template/ruin/space/thederelict): load(space (11,92,8) (/turf/open/space), 1)
[21:48:49]seedRuins(/list (/list), 16, /area/space (/area/space), /list (/list))
[21:48:49]Master (/datum/controller/master): Setup(10, 0)
````

So, this is very bad, a stack overflow runtimes kill the entire call chain.

**Comment by MrStonedOne** [comment]
Looking at `add to active` i'm actually kinda shocked we haven't seen this before.

The proc is just asking for this kind of bug to chop up.

The issue, at its core, is that open turfs are adding themselves to active before getting an air datum, so it's reverting to the behavior of adding the turf's neighbors to active.

Normally this issue an issue, because neighbors aren't calculated until after shit gets their air datum, and neighbors only include shit that can have air, so it normally doesn't recurse.

**Comment by MrStonedOne** [comment]
Status? I've been looking forward to this pr all week and so everything i see it not fixed i cry a little

**Comment by Cyberboss** [comment]
It's party night, I'll fix it when the hangover stops

File: code/controllers/subsystem/air.dm [hunk 6]
```diff
@@ -244,8 +249,18 @@ var/datum/subsystem/air/SSair
 		if(T.excited_group)
 			T.excited_group.garbage_collect()
 
+/datum/subsystem/air/proc/activate_queued_turfs()
+	if(LAZYLEN(to_activate))
+		for(var/T in to_activate)
+			add_to_active(T, to_activate[T])
+	to_activate = null
```
Comment (MrStonedOne) [inline code/controllers/subsystem/air.dm]: This isn't actually needed because turfs check if they need to be in the active list when ssair runs setup_allturfs()

Comment (MrStonedOne) [inline code/controllers/subsystem/air.dm]: also, the bug is that turfs should not be checking for `atmos_adjacent_turfs` before ssair init (because it's done in ssair init) and that non-open turfs are ending up in the `atmos_adjacent_turfs` list.

**Comment by MrStonedOne** [comment]
I'll re-download this branch and give it another test run on my computer once i'm done debugging timers.

**Comment by optimumtact** [comment]
@MrStonedOne let me know on this one, It looks sensible to me

**Comment by MrStonedOne** [comment]
```
[02:25:20]Cannot create objects of type null.
[02:25:20]proc name: SpawnFlora (/turf/open/floor/plating/asteroid/airless/cave/proc/SpawnFlora)
[02:25:20]  source file: asteroid.dm,301
[02:25:20]  usr: null
[02:25:20]  src: the asteroid sand (188,68,5) (/turf/open/floor/plating/asteroid/airless/cave/volcanic/has_data)
[02:25:20]  call stack:
[02:25:20]the asteroid sand (188,68,5) (/turf/open/floor/plating/asteroid/airless/cave/volcanic/has_data): SpawnFlora(the rock (188,82,5) (/turf/closed/mineral/random/volcanic))
[02:25:20]the asteroid sand (188,68,5) (/turf/open/floor/plating/asteroid/airless/cave/volcanic/has_data): SpawnFloor(the rock (188,82,5) (/turf/closed/mineral/random/volcanic))
[02:25:20]the asteroid sand (188,68,5) (/turf/open/floor/plating/asteroid/airless/cave/volcanic/has_data): make tunnel(1)
[02:25:20]the asteroid sand (188,68,5) (/turf/open/floor/plating/asteroid/airless/cave/volcanic/has_data): produce tunnel from data(14, 5)
[02:25:20]the asteroid sand (174,54,5) (/turf/open/floor/plating/asteroid/airless/cave/volcanic): make tunnel(5)
[02:25:20]the asteroid sand (174,54,5) (/turf/open/floor/plating/asteroid/airless/cave/volcanic): produce tunnel from data(null, -1)
[02:25:20]the asteroid sand (174,54,5) (/turf/open/floor/plating/asteroid/airless/cave/volcanic): Initialize(1)
[02:25:20]Objects (/datum/subsystem/objects): Initialize(375087)
[02:25:20]Master (/datum/controller/master): Setup(10, 0)
```

**Comment by MrStonedOne** [comment]
There are a fuck ton of these.

**Comment by Cyberboss** [comment]
I know what this problem is. The floors are ChangeTurf-ing in the Initialize call. So they expect the new turf to have the data. But Initialize isn't set up to handle new objs created during it yet. That's what #23274 is for. Merge that first and this should fix itself.

**Comment by Cyberboss** [comment]
This'll also fix the problem with turfs getting the `initialized twice` runtime when ChangeTurf-ing

**Comment by Cyberboss** [comment]
DNM

File: code/game/turfs/turf.dm [hunk 7]
```diff
@@ -183,6 +183,8 @@
 	var/old_ex_id = explosion_id
 
 	Destroy()	//❄
+	if(initialized)
+		world.log << "Fuck your shit ass cunt"
```
Comment (lzimann) [inline code/game/turfs/turf.dm]: This confused me a lot when I woke up and someone said there was a bunch of "Fuck your shit ass cunt" in the runtime logs.

Comment (kevinz000) [inline code/game/turfs/turf.dm]: CYYBERFUCKINGBOSS!!!!!!

**Comment by Cyberboss** [comment]
Coding at 2AM. Not even once

**Comment by Cyberboss** [comment]
@lzimann  retestmerge

**Comment by Cyberboss** [comment]
OK, so, the bugs are all cleaned up. However, I just realized that I didn't take care of `/turf/open/New` which presents it's own problems :/

**Comment by Cyberboss** [comment]
Current stats (Note: this is my slow ass computer, hence the massive times): 

World init: 29.987s
SSobj: 98.7s
Total: 128.687s

With /turf/open/New removal:

World init: 13.675s
SSobj: 86.2
Total: 99.875s

Gonna go ahead and add that

File: code/modules/atmospherics/environmental/LINDA_turf_tile.dm [hunk 8]
```diff
@@ -79,6 +79,8 @@
 	return GM
 
 /turf/open/return_air()
+	if(!initialized)
+		Initialize()
 	return air
```
Comment (Cyberboss) [inline code/modules/atmospherics/environmental/LINDA_turf_tile.dm]: @MrStonedOne

**Comment by Cyberboss** [comment]
Updated stats from my good PC:

World init: 9.365s
SSobj: 38.2s
Total : 47.565s

Wew

**Comment by RemieRichards** [comment]
@Cyberboss do you not understand the ```/static``` keyword? ->thatlastcommit

**Comment by Cyberboss** [comment]
I'm assuming it's the same as C++? One instance across all objects?

**Comment by RemieRichards** [comment]
@Cyberboss Yes, in DM it also abstracts out the assignment of the variable to some spook-zone where code doesn't actually run but the assignment still happens, Confused? see this:

```
/mob/verb/test_static_keyword_return_proc()
	set category = "Static"
	var/static/list/return_static = return_me_a_list()
	world << "\ref[return_static]"
	
/proc/return_me_a_list()
	var/static/i = 0
	world << "[++i] runs"
	world << "spook"
	return list()
```

You'll note a few things when you run this multiple times:
1. the ref of return_static is always the same, so it's not getting re-assigned a new list
and
2. the two world statements never happen (and yet, the ```return list()``` did?)

Since typecacheof is also a proc, it obeys the same rules as ```return_me_a_list()```, and as such, the assignment only ever happens once, even for multiple mobs running that same section of code, it's as if it's been removed after the first assignment, and only important (those parts that control the value) parts of the assignment are ran

This means that your "improvement" to the ```typecacheof()``` calls does... nothing but adding the overhead of a conditional branch

tl;dr remove that commit

**Comment by MrStonedOne** [comment]
Remie, do you even lift?

```
/turf/open/floor/plating/lava/Initialize()
	..()
	safeties_typecache = typecacheof(safeties_typecache)
```

The actual conversion isn't part of the static proc's initialization

Cyber's commit was correct.

**Comment by RemieRichards** [comment]
@MrStonedOne  if you weren't busy yesterday when I was pinging you, I could have shown you this:

```
/mob/var/static/list/def_static = list()
/mob/verb/test_static_def()
	set category = "Static"
	def_static = return_me_a_list()
	world << "\ref[def_static]"

/proc/return_me_a_list()
	var/static/i = 0
	world << "[++i] runs"
	world << "spook"
	return list()

```
The ref of def_static flips between 2 different IDs, but the rest of the proc still runs, in my test world (not relevant but) the refs are [0xf00000a] and [0xf000004], any ideas?

File: code/game/turfs/turf.dm [hunk 9]
```diff
@@ -21,10 +21,10 @@
...
 	var/explosion_id = 0
-	
+
 	var/list/decals
 
-/turf/New()
+/turf/Initialize()
 	..()
 
 	levelupdate()
```
Comment (MrStonedOne) [inline code/game/turfs/turf.dm]: The queue_smooth inside this proc accounts for 1/4 of ssobject's initialization.

since SSicon_smooth initializes later on and already handles smoothing zlevels 1 and 2 on initialize(), it might be worth the snowflake to have queue_smooth do nothing if the turf would be smooth'ed when SSicon_smooth initializes (and SSicon_smooth hasn't initialized yet.)

Overall I love this pr! Mainly because now we can see and easily profile the cost our turfs were having on world init.

When are you going to make all objs use initialize? =P

Comment (MrStonedOne) [inline code/game/turfs/turf.dm]: On second thought, this should be done in another PR. I wonder if smoothing the zlevel is even wise for it's initialize since everything forces a smooth queuing it should be fine to just remove the zlevel smoothing bit from SSicon_smooth's initialize.

Then you can make SSicon_smooth much much faster by adding a flag to smooth for if something is already queued to smooth, and check that rather then do an associated add.

Comment (MrStonedOne) [inline code/game/turfs/turf.dm]: Ya, I just confirmed that the smoothing of the 1st two zlevels as a seperate step in SSicon_smooth's initialize() is no longer needed (potentially because of this pr since i haven't bothered to test on master).

Also, changing queue_smooth()'s `SSicon_smooth.smooth_queue[A] = A` to `SSicon_smooth.smooth_queue += A` and tracking it by adding a bit flag to `atom.smooth` to avoid dupes removes all of the overhead of queue_smooth, going down from 5 seconds to 0.2.

I'll just pr that last bit as a pr

**Comment by MrStonedOne** [comment]
Test merged on both servers

**Comment by lzimann** [comment]
Pretty sure this PR caused [this](https://file.house/7bX2.txt), it's the only test merged that touches air somehow.

**Comment by Cyberboss** [comment]
Jesus what. @MrStonedOne and yes, my goal is to eventually move us entirely
from New to Initialize for atoms.

**Comment by MrStonedOne** [comment]
https://tgstation13.org/parsed-logs/sybil/runtimes/runtime-2017-02-04-condensed.txt

**Comment by Cyberboss** [comment]
Not again...

**Comment by optimumtact** [comment]
is this ready, is #23552 ready, is #23278 ready?

**Comment by Cyberboss** [comment]
#23278 is gone methinks. No to everything else.

**Comment by Cyberboss** [comment]
Fine*

File: code/controllers/subsystem/air.dm [hunk 10]
```diff
@@ -253,10 +255,25 @@ var/datum/subsystem/air/SSair
 			currentrun |= T
 		if(blockchanges && T.excited_group)
 			T.excited_group.garbage_collect()
-	else
+	else if(T.initialized)
 		for(var/turf/S in T.atmos_adjacent_turfs)
 			add_to_active(S)
+	else if(map_loading)
+		if(queued_for_activation)
+			queued_for_activation |= T
```
Comment (RemieRichards) [inline code/controllers/subsystem/air.dm]: ```queue_for_activation[T] = T``` is faster than ```queue_for_activation |= T```

**Comment by RemieRichards** [comment]
@MrStonedOne surprised you weren't interested in the whole "hey this static var has 2 refs" thing.

**Comment by Cyberboss** [comment]
Testmerge seems fine so far. @optimumtact if nothing comes up in 24hrs you can merge this

**Comment by lzimann** [comment]
Minerals on lavaland doesn't seem to be spawning with this.

**Comment by Cyberboss** [comment]
Fuuuuuuuuuuuuuuuuuuuuuuuuuuuuck

**Comment by Cyberboss** [comment]
OK, I've managed to prove that its one of the ruins that's fucking with it and firing off that infinite loop bug. 

I just don't know which ruin...

... Fuck ruins

**Comment by Cyberboss** [comment]
Same dealio as before, 24h from now. No complaints = merge
